File: Unique_binary_search_trees/solution_dp.py
class Solution:
    # @return an integer
    def numTrees(self, n):
        #dynamic programming
        #for each i, caculate how many unique bst are there when i is root
        #store them in res(when caculate i+1, we can use the result of i)
        ###################
        res = [1,1]
        #caculate how many unique bst are there when i is root
        for i in range(2,n+1):
            sum = 0
            #sum is the number of bst when i is root
            #enumerate different left/right subtree(use j as control variable)
            for j in range (0,i):
                sum += res[j]*res[i-j-1]
            res.append(sum)
        return res[n]

File: Unique_binary_search_trees/solution_dp.cpp
class Solution {
public:
    int numTrees(int n) {
        //dp
        vector<int> res;
        res.push_back(1);
        res.push_back(1);
        if(n==0) return res[0];
        if(n==1) return res[1];
        for(int i=2; i<=n; i++){
            int sum = 0;
            for(int j=0; j<i; j++){
                sum = sum+res[j]*res[i-j-1];
            }
            res.push_back(sum);
        }
        return res[n];
    }
};